numpy/random/mtrand/mtrand.pyx

```diff
@@ -4602,6 +4602,141 @@ cdef class RandomState:
+
+        Uses the following property for computation: for each dimension,
+        draw a random sample y_i from a standard gamma generator of shape
+        `alpha_i`, then
+        :math:`X = \\frac{1}{\\sum_{i=1}^k{y_i}} (y_1, \\ldots, y_n)` is
+        Dirichlet distributed.
+
+        References
+        ----------
+        .. [1] David McKay, "Information Theory, Inference and Learning
+               Algorithms," chapter 23,
+               http://www.inference.phy.cam.ac.uk/mackay/
+        .. [2] Wikipedia, "Dirichlet distribution",
+               http://en.wikipedia.org/wiki/Dirichlet_distribution
+
+        Examples
+        --------
+        Taking an example cited in Wikipedia, this distribution can be used if
+        one wanted to cut strings (each of initial length 1.0) into K pieces
+        with different lengths, where each piece had, on average, a designated
+        average length, but allowing some variation in the relative sizes of
+        the pieces.
+
+        >>> s = np.random.dirichlet((10, 5, 3), 20).transpose()
+
+        >>> plt.barh(range(20), s[0])
+        >>> plt.barh(range(20), s[1], left=s[0], color='g')
+        >>> plt.barh(range(20), s[2], left=s[0]+s[1], color='r')
+        >>> plt.title("Lengths of Strings")
+
+        """
+
+        #=================
+        # Pure python algo
+        #=================
+        #alpha   = N.atleast_1d(alpha)
+        #k       = alpha.size
+
+        #if n == 1:
+        #    val = N.zeros(k)
+        #    for i in range(k):
+        #        val[i]   = sgamma(alpha[i], n)
+        #    val /= N.sum(val)
+        #else:
+        #    val = N.zeros((k, n))
+        #    for i in range(k):
+        #        val[i]   = sgamma(alpha[i], n)
+        #    val /= N.sum(val, axis = 0)
+        #    val = val.T
+
+        #return val
+
+        cdef npy_intp   k
+        cdef npy_intp   totsize
+        cdef ndarray    alpha_arr, val_arr
+        cdef double     *alpha_data
+        cdef double     *val_data
+        cdef npy_intp   i, j
+        cdef double     acc, invacc
+
+        cdef double alpha_max
+        cdef double alpha_sum
+
+        k           = len(alpha)
+        alpha_arr   = <ndarray>PyArray_ContiguousFromObject(alpha, NPY_DOUBLE, 1, 1)
+        alpha_data  = <double*>PyArray_DATA(alpha_arr)
+
+        shape = _shape_from_size(size, k)
+
+        diric   = np.zeros(shape, np.float64)
+        val_arr = <ndarray>diric
+        val_data= <double*>PyArray_DATA(val_arr)        
+
+        i = 0
+        totsize = PyArray_SIZE(val_arr)
+        alpha_max = alpha_arr.max()
+
+        if alpha_max < 1.0:
+            alpha_sum = alpha_arr.sum()
+            """Generation via beta-RVs"""            
+            with self.lock, nogil:
+                while i < totsize:
+                    acc = 1.0
+                    invacc = alpha_sum
+                    for j from 0<= j < k-1:
+                        invacc -= alpha_data[j]
```

Comment:
This shouldn't stop the technical discussion, but FTR I'm -1 on merging
this until we have some solution that doesn't expose a public dirichlet2
method, for reasons discussed in #5851. In case anyone starts getting any
itchy merge fingers *cough*@charris*cough* ;-).
On May 13, 2015 1:21 PM, "argriffing" <notifications@github.com> wrote:

> In numpy/random/mtrand/mtrand.pyx
> <https://github.com/numpy/numpy/pull/5872#discussion_r30271641>:
>
> > +        val_arr = <ndarray>diric
> > +        val_data= <double*>PyArray_DATA(val_arr)
> > +
> > +        i = 0
> > +        totsize = PyArray_SIZE(val_arr)
> > +        alpha_max = alpha_arr.max()
> > +
> > +        if alpha_max < 1.0:
> > +            alpha_sum = alpha_arr.sum()
> > +            """Generation via beta-RVs"""
> > +            with self.lock, nogil:
> > +                while i < totsize:
> > +                    acc = 1.0
> > +                    invacc = alpha_sum
> > +                    for j from 0<= j < k-1:
> > +                        invacc -= alpha_data[j]
>
> If there is a failure with only two elements in the array then I'm not
> understanding something because the 2-element case is the same as the beta
> distribution. I'm assuming we can sample from a beta distribution.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/numpy/numpy/pull/5872/files#r30271641>.
>